articles/202503151400-search-engine.md

```diff
@@ -0,0 +1,363 @@
+
+# Building a Search Engine from Scratch in Rust
+
+In January of this year, I was hired by a company dedicated to keeping client data private. To achieve this, they encrypt as much data as possible, making efficient search functionality a significant challenge.
+
+## The Challenge
+
+The core issue was simple: **How do you allow users to search through their encrypted data while ensuring it remains private?**
+Standard search solutions were not an option, as they typically require decrypting data on the server—something we couldn’t allow.
+
+## Key Requirements
+
+- **Privacy-first indexing**: Once data is indexed, it should remain encrypted on disk, preventing content extraction through index analysis.
+- **On-device search**: Since decrypting data on the server was not viable and homomorphic search is not yet practical, all indexing and searching had to happen directly on the user's device.
```

Comment:
@jdrouet suggestion: what do you think to replace `homomorphic search` by `[Searchable symmetric encryption](https://en.wikipedia.org/wiki/Searchable_symmetric_encryption)` ?